src/cmdstan/command_helper.hpp

```diff
@@ -785,17 +785,24 @@ void init_callbacks(
       ofs->precision(sig_figs);
     sample_writers.emplace_back(std::move(ofs), "# ");
   }
+
+  diag_json_writers.reserve(num_chains);
+  diag_csv_writers.reserve(num_chains);
+  // create no-op writers by default
+  for (int i = 0; i < num_chains; ++i) {
+    diag_json_writers.emplace_back(
+        stan::callbacks::json_writer<std::ofstream>());
+  }
+  for (int i = 0; i < num_chains; ++i) {
+    diag_csv_writers.emplace_back(nullptr, "# ");
+  }
+  // create json, csv writers as needed.
   std::string diagnostic_file
       = get_arg_val<string_argument>(parser, "output", "diagnostic_file");
-  if (user_method->arg("pathfinder")) {
-    diag_json_writers.reserve(num_chains);
-    if (diagnostic_file.empty()) {
-      for (int i = 0; i < num_chains; ++i) {
-        diag_json_writers.emplace_back(
-            stan::callbacks::json_writer<std::ofstream>());
-      }
-    } else {
-      std::vector<std::string> diag_filenames;
+  if (!diagnostic_file.empty()) {
+    std::vector<std::string> diag_filenames;
+    if (user_method->arg("pathfinder")) {
+      diag_json_writers.resize(0);
```

Comment:
Basic C++ question: is there a reason to prefer `.resize(0)` to `.clear()`?

Reply:
maybe?  discussion inconclusive:  https://stackoverflow.com/questions/2738967/vector-clear-vs-resize

Reply:
It sounds like in C++11 and later clear is not allowed to change the capacity, so either seems fine.

Reply:
I believe the current test failures are actually caused by the `.remove(0)`, since `remove` requires default-constructable types. Changing it to clear() resolves the issue for me locally.